File: react-canvas/src/game/entities/Bullet.ts
import type { Vector2D } from '../types';
import {
  CANVAS_WIDTH,
  CANVAS_HEIGHT,
  BULLET_SPEED,
  BULLET_LIFETIME,
  BULLET_RADIUS,
} from '../constants';

export class Bullet {
  position: Vector2D;
  velocity: Vector2D;
  radius: number;
  lifetime: number;
  alive: boolean;

  constructor(position: Vector2D, angle: number) {
    this.position = { ...position };
    this.velocity = {
      x: Math.cos(angle) * BULLET_SPEED,
      y: Math.sin(angle) * BULLET_SPEED,
    };
    this.radius = BULLET_RADIUS;
    this.lifetime = BULLET_LIFETIME;
    this.alive = true;
  }

  update(dt: number): void {
    if (!this.alive) return;

    this.lifetime -= dt;
    if (this.lifetime <= 0) {
      this.alive = false;
      return;
    }

    // Position
    this.position.x += this.velocity.x * dt;
    this.position.y += this.velocity.y * dt;

    // Screen wrapping
    if (this.position.x < 0) this.position.x = CANVAS_WIDTH;
    if (this.position.x > CANVAS_WIDTH) this.position.x = 0;
    if (this.position.y < 0) this.position.y = CANVAS_HEIGHT;
    if (this.position.y > CANVAS_HEIGHT) this.position.y = 0;
  }
}

File: python-pygame/game/entities/bullet.py
"""Bullet entity for the Asteroids game."""

import math
from game.constants import (
    SCREEN_WIDTH, SCREEN_HEIGHT, BULLET_SPEED, BULLET_LIFETIME, BULLET_RADIUS
)


class Bullet:
    """A bullet fired from the ship."""

    def __init__(self, x, y, angle):
        rad = math.radians(angle)
        self.x = x
        self.y = y
        self.vx = math.cos(rad) * BULLET_SPEED
        self.vy = math.sin(rad) * BULLET_SPEED
        self.lifetime = BULLET_LIFETIME
        self.radius = BULLET_RADIUS
        self.alive = True

    def update(self, dt):
        """Update bullet position and lifetime."""
        self.x += self.vx * dt
        self.y += self.vy * dt

        # Screen wrapping
        self.x = self.x % SCREEN_WIDTH
        self.y = self.y % SCREEN_HEIGHT

        # Lifetime countdown
        self.lifetime -= dt
        if self.lifetime <= 0:
            self.alive = False
